adapters/huaweiads/huaweiads.go

```diff
@@ -788,39 +778,60 @@ func convertCountryCode(country string) (out string) {
 	return defaultCountryName
 }
 
+func getCountryCodeFromMCC(MCC string) (out string) {
+	var countryMCC = strings.Split(MCC, "-")[0]
+	intVar, err := strconv.Atoi(countryMCC)
+
+	if err != nil {
+		return defaultCountryName
+	} else {
+		if result, found := MccList[intVar]; found {
+			return strings.ToUpper(result)
+		} else {
+			return defaultCountryName
+		}
+	}
+}
+
 // getDeviceID include oaid gaid imei. In prebid mobile, use TargetingParams.addUserData("imei", "imei-test");
 // When ifa: gaid exists, other device id can be passed by TargetingParams.addUserData("oaid", "oaid-test");
-// When getGaidFromDeviceIFA success, getDeviceIDFromUserExt failed, no error will be reported.
-func getDeviceIDFromUserExt(device *device, isGetGaidSuccess bool, openRTBRequest *openrtb2.BidRequest) (err error) {
+func getDeviceIDFromUserExt(device *device, openRTBRequest *openrtb2.BidRequest) (err error) {
+	var userObjExist = true
 	if openRTBRequest.User == nil || openRTBRequest.User.Ext == nil {
-		if isGetGaidSuccess {
-			return nil
-		}
-		return errors.New("get gaid from openrtb Device.IFA failed, and get device id failed: openRTBRequest.User.Ext is nil.")
+		userObjExist = false
 	}
-	var extUserDataHuaweiAds openrtb_ext.ExtUserDataHuaweiAds
-	if err := json.Unmarshal(openRTBRequest.User.Ext, &extUserDataHuaweiAds); err != nil {
-		if isGetGaidSuccess {
-			return nil
+	if userObjExist {
+		var extUserDataHuaweiAds openrtb_ext.ExtUserDataHuaweiAds
+		if err := json.Unmarshal(openRTBRequest.User.Ext, &extUserDataHuaweiAds); err != nil {
+			return errors.New("get gaid from openrtb Device.IFA failed, and get device id failed: Unmarshal openRTBRequest.User.Ext -> extUserDataHuaweiAds. Error: " + err.Error())
+		}
+
+		var deviceId = extUserDataHuaweiAds.Data
+		isValidDeviceId := false
+
+		if len(deviceId.Oaid) > 0 {
+			device.Oaid = deviceId.Oaid[0]
+			isValidDeviceId = true
+		}
+		if len(deviceId.Gaid) > 0 {
+			device.Gaid = deviceId.Gaid[0]
+			isValidDeviceId = true
+		}
+		if len(deviceId.Imei) > 0 {
+			device.Imei = deviceId.Imei[0]
+			isValidDeviceId = true
```

Comment:
Is there any way to trigger/cover these lines of code

Reply:
banner6_imei.json file is added to cover deviceId.Imei

Reply:
@GenarC I still see this line uncovered when I run the tests in the huaweiads package